### ISSUE TITLE
Bazel C/C++ toolchain autodetection fails with newest version of Visual Studio on Windows (17.6.2) due to new directory layout

### ISSUE THREAD

**Issue by redsun82** [bazelbuild/bazel#18592]
### Description of the bug:

When running C/C++ builds on Windows with autodetection where the most recent version of Visual Studio Enterprise is installed (version 17.6.2), the build fails with
```
Auto-Configuration Error: Cannot find VCVARSALL.BAT script under C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC
```

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

* Install latest Visual Studio (I did not try with the build tools version though), at least 17.6.2
* Compile any dummy C/C++ project not using any kind of custom toolchain (for example, `//examples/cpp:hello-world` in the bazel repo itself)

### Which operating system are you running Bazel on?

Windows

### What is the output of `bazel info release`?

release 6.2.1

### Any other information, logs, or outputs that you want to share?

By debugging I've traced back this behaviour to [this _is_vs_2017_or_2019 function](https://github.com/bazelbuild/bazel/blob/030d94cf4cb23c3e4581141774a1b3d8cf828a85/tools/cpp/windows_cc_configure.bzl#L257-L267)

Apart from the misname (it's supposed to actually check whether the release year is greater than 2017, including 2022), the problem is that [a recent addition to VS](https://devblogs.microsoft.com/cppblog/vcpkg-is-now-included-with-visual-studio/) has added a fourth `vcpkg` directory in the `VC` directory, that throws the check off.

The check could be updated to include the presence of this `vcpkg` directory, but I wonder whether there wouldn't be a more solid alternative robust to similar changes in the future. Maybe
```bazel
def _is_vs_2017_or_later(repository_ctx, vc_path):
    # example path C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC
    release_year = int(repository_ctx.path(vc_path).dirname.dirname.basename)
    return release_year >= 2017
```
could work better? I can whip up a PR if this seems like a good idea.

**Comment by redsun82** [comment]
Here's my rough shot at a fix, though I'm not entirely sure how to add tests for it

https://github.com/bazelbuild/bazel/pull/18608

**Comment by redsun82** [comment]
I've noticed no prior changes to the autoconfiguration included any kind of tests, so I'm marking [my PR](https://github.com/bazelbuild/bazel/pull/18608) as ready for review, after trying it out locally.

I've also noticed that all Windows CI checks currently use VS 2019. It might probably be a good idea to add or switch to VS 2022.

As a temporary workaround that I used to compile bazel itself, one can go as admin into the `VC` directory (for example `C:\Program Files\Microsoft Visual Studio\2022\Community\VC` and `move vcpkg ..`. I did not notice anything amiss in compilation then, but YMMV. I've then undone the change with `move ../vcpkg .`.

**Comment by Mizux** [comment]
beware that Visual Studio 2022 **Community** and **Enterprise** are not the same (FYI github runner and this user is using `Enterprise` version)

**Comment by redsun82** [comment]
Yes, but both actually exhibit the same behaviour and work with the same workaround of moving the `vcpkg` directory out of the way. I could see after dealing both with a virtual machine at work (with the `Enterprise` edition) and my personal machine at home (with the `Community` edition). I did not look at the `Professional` edition but I'd expect the same. `BuildTools` might not have that directory, or it's just that I did not have a recent version enough. In any case the proposed PR deals gracefully with the presence or absence of that directory in any case.

**Comment by lblk** [comment]
还要多久

**Comment by redsun82** [comment]
FYI, I think there is another workaround:

https://github.com/bazelbuild/bazel/pull/18608#issuecomment-1598294872

**Comment by redsun82** [comment]
and if you trust me enough, while you wait for the fix to get to official bazel, you can download a patched build here:

https://github.com/redsun82/bazel/releases/tag/vs-2022-fix-0.1

if you don't trust me, well, you can build from the source provided there 😉

**Comment by lblk** [comment]
i compile an repo :https://github.com/alibaba/async_simple.git by using  your release version on vs2022, it saves me at least a week. thank u so much
